README.md

```diff
@@ -1 +1,218 @@
-# node-patterns-concurrency
\ No newline at end of file
+# Concurrency and Other Async Patterns
+
+## Overview
+
+Asynchronous code is more efficient than synchronous, but to write good async code you need to know at least a few Node patterns.
+
+This lesson will cover the concurrency and error handling patterns.
+
+## Objectives
+
+1. Learn async error handling instead of try/catch
+1. Demonstrate concurrency with async for-loop
+
+## Sync Error Handling
+
+In traditional synchronous code, you would `throw` an error and catch it later up in the call stack... or not catch it (if you just writing a class, whatever uses it will need to catch the error). This is an illustration:
+
+```js
+var badJson = '{a: "1"}'
+try {
+  var obj = JSON.parse(badJson)
+} catch (error) {
+  console.error(error)
+}
+```  
+
+If you run this code, you'll see "[SyntaxError: Unexpected token a]" because `badJson` is a bad JSON object (no double quotes around property name `a`). The script didn't crash. Instead, `try/catch` saved it. If you had this snippet in a large application, the app would continue to work just fine albeit without the parsed object.
+
+For this reason, in synchronous code you can `throw` an error as long as you have `try/catch` to handle it:
+
+```js
+try {
+  throw new Error('BOOM!')
+} catch (error) {
+  console.error(error)
+}
+```
+
+Now, create a file `async-throw.js` and type the code which has `throw` inside of the asynchronous callback:
+
+```js
+try {
+  setTimeout(function(){  throw new Error('BOOM!')}, 0)
+} catch (error) {
+  console.error(error)
+}
+```
+
+The script will crash. This is what you will see:
+
+```
+KA...
+/Users/azat/Documents/Code/learn-co/node-patterns-concurrency/async-throw.js:2
+  setTimeout(function(){  throw new Error('BOOM!')}, 0)
+                          ^
+
+Error: BOOM!
+    at null._onTimeout (/Users/azat/Documents/Code/learn-co/node-patterns-concurrency/async-throw.js:2:33)
+    at Timer.listOnTimeout (timers.js:92:15)
+```
+
+
+Note: When you `setTimeout` with 0, it will schedule the callback at 0 milliseconds, but on the next tick (same as iteration or cycle) of the event loop. For this reason, Node will finish with all the statements in this file first and only after that on the next cycle, it will fire up the callback.
+
+
+The "KA..." message is shown first, then the delayed callback throws an error which kills the program. When we don't have asynchronous code, `try/catch` works. However, with asynchronous code `try/catch` is useless unless you put it inside of the callback. Why is that?
+
+If we go deeper under the hood, callbacks are scheduled by the event loop. They lose context of the  environment in which they were defined (and lose `try/catch` which is around them). In a way, future code (callbacks) and current code (`try/catch`) become separated by the Node's non-blocking I/O and event loop mechanisms. 
+
+What should we do?
+
+## Async Error Handling
```

Comment:
I thought we covered this in the previous lesson (or one very close to this)? Why do we have a whole half of a lesson for async error handling?